Added check to see if folder is a valid node module

pezza3434 opened the pull request:
Added check to see if a folder is a node module as per require specs.

[https://nodejs.org/api/modules.html#modules_folders_as_modules](https://nodejs.org/api/modules.html#modules_folders_as_modules)

@aseemk

pezza3434 committed:
Added check to see if folder is a valid node module

AlexMeah commented:
@aseemk This is really useful, and the expected behaviour would it be possible to get this merged?

aseemk commented:
Sorry I missed this; thanks for the nudge.

Interesting feature. One thing I'm unsure of is how this should relate to `recurse`. Currently, you're doing this as the default behavior. Is that best? Or is that counterintuitive / violates least surprise?

Also, would you mind adding this to the readme in the PR? Doesn't have to be perfect; it'd just be good to see how you would communicate this feature. Thanks! (And thanks for already adding tests.)

aseemk commented on index.js:
```diff
@@ -68,6 +68,17 @@ module.exports = function requireDir(dir, opts) {
+                if (modulePath && !opts.recurse) {
+                    filesMinusDirs[file + Path.extname(modulePath)] = modulePath;
+                }
+
```
(Assuming for now we keep your design that this should be the default behavior) Should this happen only if `opts.recurse` *isn't* specified?

Put another way: what should happen when `opts.recurse` *is* specified? Currently, we'd `require` the dir normally, and then that would get overwritten on recurse, is that right? So harmless, but inefficient?

Can we add a test case for what should happen when `opts.recurse` *is* specified and a (sub)directory is a module? Thanks!

aseemk commented on index.js:
```diff
@@ -100,6 +111,7 @@ module.exports = function requireDir(dir, opts) {
 
             // if a file exists with this extension, we'll require() it:
             var file = base + ext;
+
```
Accidental whitespace change I'm guessing.

aseemk commented on index.js:
```diff
@@ -134,6 +146,14 @@ module.exports = function requireDir(dir, opts) {
+    try {
+        return require.resolve(path);
+    } catch (_) {
+        return false;
```
Minor: let's `catch (err)` and `return null` instead of `false`.

aseemk commented on test/directoryAsModule/b/index.js:
```diff
@@ -0,0 +1 @@
+module.exports = 'b'
```
Can we add a second file to this directory, to explicitly test (and convey) that it isn't getting `require`d?

pezza3434 replied:
Done :)

aseemk commented on test/directoryAsModule.js:
```diff
@@ -0,0 +1,8 @@
+    a: 'a', b: 'b'
+});
+
+console.log('Automatic Index tests passed')
```
Minor: let's have this `console.log` message be consistent with the feature & test name: "Directory as module tests passed".

aseemk commented on test/directoryAsModule.js:
```diff
@@ -0,0 +1,8 @@
+var requireDir = require('..');
+
+assert.deepEqual(requireDir('./directoryAsModule'), {
+    a: 'a', b: 'b'
```
Minor: let's keep to the code style of having these object keys be on their own lines.

aseemk commented:
Good idea btw and thanks for the PR!

pezza3434 committed:
Refactored directory as a module functionality

pezza3434 committed:
Added readme about directory as modules

pezza3434 commented:
Hi @aseemk,

I've changed a lot of the small things that you mentioned, I also added to the readme.

I was also unsure about the recursion functionality. I decided to omit this functionality from the recursion option for a couple of reasons:

1) It would be a major breaking change
2) The recursion options goes through the entire directory, including sub-directories and so people may want the verbosity that it exposes.
3) Looping through all the folders like that is not a 'node thing' and to resolve the folders as such would not be expected behaviour (for me anyway).

Perhaps going forwards there could be two recursion options, with and without verbosity. Verbosity would give you all the folders and contents while non verbosity would attempt to resolve each folder and treat it how we have in this pull request. 

I look forward to your thoughts on the changes.

pezza3434 committed:
Removed white space

pezza3434 commented:
@aseemk Any updates on this? Thanks!

aseemk commented:
Sorry folks, I've been unusually busy with both work and life for a sustained period of time now. I keep meaning to get to these PRs (and other projects), but don't find myself with the time + energy too often.

One thing I've done with another project is to invite one or two other people to be collaborators in the project and help maintain it. Anyone here interested in that?

I'd just ask that you be mindful of keeping this project focused, and not let it get too complex. E.g. I'm not 100% sure it's a good idea to accept all of the currently open PRs, but I haven't looked into them fully or thought deeply about it.

Apologies again for the delay, and thank you for understanding. =)

aseemk commented:
Hi there,

I'm sorry I haven't been timely in responding to these issues and PRs.

If you're interested in taking over the maintainer role for this project, please let me know.

More details: https://github.com/aseemk/requireDir/pull/31#issuecomment-255744987

Cheers,
Aseem

contra commented:
@pezza3434 Could you rebase this? I'll try to get this merged and published ASAP. Thanks for the contribution.

Janpot commented:
It makes sense that this module follows node's cjs resolution algorithm by default. I'd like to argue that `recurse` should follow this resolution algorithm as well.
I understand that this is a breaking change and that it would break a lot of code, including my own.
Maybe it makes sense to add a new `deep` option that would recurse through folders but uses `require`'s resolution algorithm?

contra commented:
@Janpot I agree, I think this should be the default behavior (recurse or no recurse) - as for breaking code, I'll publish it as a breaking release with a note in the changelog.

contra commented:
ping @pezza3434

thealjey commented:
I need this :)
